state/modelsummaries_test.go

```diff
@@ -0,0 +1,390 @@
+	summaries, err := s.State.ModelSummariesForUser(s.Model.Owner(), false)
+	c.Assert(err, jc.ErrorIsNil)
+	names := make([]string, len(summaries))
+	for i, summary := range summaries {
+		names[i] = summary.Name
+	}
+	sort.Strings(names)
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForSuperuserWithAll(c *gc.C) {
+	s.Setup4Models(c)
+	summaries, err := s.State.ModelSummariesForUser(s.Model.Owner(), true)
+	c.Assert(err, jc.ErrorIsNil)
+	names := make([]string, len(summaries))
+	access := make(map[string]string)
+	for i, summary := range summaries {
+		names[i] = summary.Name
+		access[summary.Name] = string(summary.Access)
+	}
+	sort.Strings(names)
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv", "user1model", "user2model", "user3model"})
+	c.Check(access, gc.DeepEquals, map[string]string{
+		"shared":     "admin",
+		"testenv":    "admin",
+		"user1model": "",
+		"user2model": "",
+		"user3model": "",
+	})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser1(c *gc.C) {
+	// User1 is only added to the model they own and the shared model
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, "user1write")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user1model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser2(c *gc.C) {
+	// User1 is only added to the model they own and the shared model
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, "user2read")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user2model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForUser3(c *gc.C) {
+	s.Setup4Models(c)
+	names := s.modelNamesForUser(c, "user3admin")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user3model"})
+}
+
+func (s *ModelSummariesSuite) TestModelsForIgnoresImportingModels(c *gc.C) {
+	s.Setup4Models(c)
+	cfg := testing.CustomModelConfig(c, testing.Attrs{
+		"name": "importing",
+		"uuid": utils.MustNewUUID().String(),
+		"type": state.ModelTypeIAAS,
+	})
+	_, stImporting, err := s.State.NewModel(state.ModelArgs{
+		Type:                    state.ModelTypeIAAS,
+		CloudName:               "dummy",
+		CloudRegion:             "dummy-region",
+		Config:                  cfg,
+		Owner:                   names.NewUserTag("user1write"),
+		MigrationMode:           state.MigrationModeImporting,
+		EnvironVersion:          s.Model.EnvironVersion(),
+		StorageProviderRegistry: storage.StaticProviderRegistry{},
+	})
+	defer stImporting.Close()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Since the new model is importing, when we do the list we shouldn't see it.
+	names := s.modelNamesForUser(c, "user3admin")
+	c.Check(names, gc.DeepEquals, []string{"shared", "user3model"})
+	// Superuser doesn't see importing models, either
+	names = s.modelNamesForUser(c, s.Model.Owner().Name())
+	c.Check(names, gc.DeepEquals, []string{"shared", "testenv", "user1model", "user2model", "user3model"})
+}
+
+func (s *ModelSummariesSuite) TestContainsConfigInformation(c *gc.C) {
```

Comment:
And we'd want a test where there is no config for a model. We can follow it up once this PR lands...

Reply:
yes